pilot/pkg/status/resourcelock.go

```diff
@@ -14,47 +14,136 @@
 
 package status
 
-import "sync"
+import (
+	"context"
+	"sync"
 
-type ResourceLock struct {
+	"k8s.io/apimachinery/pkg/runtime/schema"
+)
+
+type ResourceMutex struct {
 	masterLock sync.RWMutex
-	listing    map[interface{}]*sync.Mutex
+	cache      map[lockResource]*cacheEntry
+}
+
+type cacheEntry struct {
+	// the runlock ensures only one routine is writing status for a given resource at a time
+	runLock sync.Mutex
+	// the cachelock protects writes and reads to the cache values
+	cacheLock sync.Mutex
+	// the cacheVale represents the latest version of the resource, including ResourceVersion
+	cacheVal *Resource
+	// the cacheProgress represents the latest version of the Progress
+	cacheProgress *Progress
+	countLock     sync.RWMutex
+	// track how many routines are running for this resource.  always <= 2
+	routineCount int32
+	// track if this resource has been deleted, and if so, stop writing updates for it
+	deleted bool
+}
+
+func (ce *cacheEntry) shouldRun() (result bool) {
+	ce.countLock.RLock()
+	count := ce.routineCount
+	ce.countLock.RUnlock()
+	if count < 2 {
```

Comment:
Why 2? Shouldn't we have 0 or 1 running?

Reply:
1 routine currently running, and 1 routine waiting on the lock.  This ensures that we always run one routine when a value has changed.

Reply:
isn't this changed to just run a single goroutine now? If not, I had a comment detailing how this may be possible that was marked as resolved - can we look into doing that? If its not possible no problem, but I think its worth exploring

Reply:
You're right.  Sorry, I misread your previous recommendation.  I will get this fixed, and thanks for the detailed recommendation.